### PR TITLE
Failures generation

### PR DESCRIPTION
**Pull request by mikhailmikhaylov**
Please see the sample and tell me if everything is OK or not.

### CODE REVIEW

File: failures-gen/plugin/src/main/resources/META-INF/gradle-plugins/org.spine3.tools.failuresgen.properties [hunk 1]
```diff
@@ -0,0 +1,22 @@
+
+#
+# Copyright 2015, TeamDev Ltd. All rights reserved.
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/resources/META-INF/gradle-plugins/org.spine3.tools.failuresgen.properties]: 2016

File: failures-gen-sample/src/main/proto/spine/sample/failures/deps/deps.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+option java_multiple_files = true;
+option java_generate_equals_and_hash = true;
+
+message Failure1Dep {
+    string id = 1;
+}
+
+enum Failure2DepEnum {
+    OK = 0;
+}
```
Comment (alexander-yevsyukov) [inline failures-gen-sample/src/main/proto/spine/sample/failures/deps/deps.proto]: Move EOF to new line.

File: failures-gen-sample/src/main/java/org/spine3/sample/failures/SampleEntry.java [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package org.spine3.sample.failures;
+
+public class SampleEntry {
```
Comment (alexander-yevsyukov) [inline failures-gen-sample/src/main/java/org/spine3/sample/failures/SampleEntry.java]: What's this?

File: .gitignore [hunk 4]
```diff
@@ -11,5 +11,10 @@
...
 
+# Include Gradle Wrapper into distribution
+!/gradle/wrapper/gradle-wrapper.jar
+
 *.iml
-credentials.properties
\ No newline at end of file
+credentials.properties
+
+*/generated/*
```
Comment (alexander-yevsyukov) [inline .gitignore]: Move EOF to new line.

File: failures-gen-sample/build.gradle [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+sourceSets {
+    main {
+        proto.srcDirs = ["$projectDir/src/main/proto"]
+        java.srcDirs = ["$projectDir/generated/main/java",
+                        "$projectDir/generated/main/spine",
+                        "$projectDir/src/main/java"]
+        resources.srcDirs = ["$projectDir/generated/main/resources"]
+    }
+}
```
Comment (alexander-yevsyukov) [inline failures-gen-sample/build.gradle]: EOF

File: failures-gen/build.gradle [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    compile "com.google.protobuf:protobuf-java:3.0.0-beta-2"
+}
+
+sourceSets {
+    main {
+        resources.srcDirs += files("plugin/src/main/resources")
+        groovy.srcDirs = ["plugin/src/main/groovy"]
+    }
+}
```
Comment (alexander-yevsyukov) [inline failures-gen/build.gradle]: EOF

File: failures-gen-sample/src/main/proto/spine/sample/failures/failures.proto [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    spine.sample.failures.deps.Failure1Dep id = 1;
+}
+
+message Failure2 {
+    spine.sample.failures.deps.Failure2DepEnum id = 1;
+}
+
+message Failure3 {
+    string id = 1;
+}
```
Comment (alexander-yevsyukov) [inline failures-gen-sample/src/main/proto/spine/sample/failures/failures.proto]: EOF

File: failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy [hunk 8]
```diff
@@ -0,0 +1,146 @@
+package org.spine3.gradle
+
+import com.google.protobuf.Descriptors.FieldDescriptor.JavaType;
+import com.google.protobuf.Descriptors;
+
+/* package */
+
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy]: No need for empty lines after 'package'.

File: failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        writePackage(osw);
+        writeImports(osw);
+        writeClassName(osw);
+        writeConstructor(osw);
+        writeGetFailure(osw);
+        writeEnding(osw);
+
+        osw.flush();
+        osw.close();
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy]: Can we use try block with Autoclosable here?

File: failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    private void writeImports(OutputStreamWriter writer) {
+
+        Set<String> dependencies = new HashSet<>();
+        fields = new LinkedHashMap<>();
+
+        for (Descriptors.FieldDescriptor field : failureDescriptor.getFields()) {
+            if (field.javaType == JavaType.MESSAGE) {
+                def dependencyFileName = field.messageType.file.fullName;
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy]: How about 'switch' here?

File: failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailuresGenPlugin.groovy [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+        File failuresInputJava = seekForFile(target, javaInputPath, failuresJavaName);
+        if (failuresInputJava == null || !failuresInputJava.exists()) {
+            log.error("No failures found.");
+        }
+        List<String> lines = failuresInputJava.readLines();
+
+        String fPackage = null;
+        String fSampleFailure = null;
+
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailuresGenPlugin.groovy]: Why 'sample'?

File: failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+            case JavaType.INT:
+            case JavaType.LONG:
+            case JavaType.FLOAT:
+            case JavaType.DOUBLE:
+            case JavaType.BOOLEAN:
+                return type.name().toLowerCase();
+                break;
+
+            default: return null;
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailureWriter.groovy]: Can we annotate method as @Nullable?

File: failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailuresGenPlugin.groovy [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        URL[] args = [fileUrl] as URL[];
+
+        URLClassLoader classLoader = new URLClassLoader(args, this.classLoader);
+
+        String className = metadata.failuresPackage + ".Failures\$" + metadata.sampleFailureName;
+
+        Class failuresClass = classLoader.loadClass(className);
+
+        return failuresClass
```
Comment (alexander-yevsyukov) [inline failures-gen/plugin/src/main/groovy/org/spine3/gradle/FailuresGenPlugin.groovy]: I think we should be consistent with usage of semicolons.

**Comment by mikhailmikhaylov** [comment]
@alexander-yevsyukov 
PTAL

**Comment by alexander-yevsyukov** [comment]
With withxxx syntax it looks better. Please satisfy Travis checks, and it would LGTM.

**Comment by alexander-yevsyukov** [comment]
Please also see me comment on `anyFailure` variable.

**Comment by mikhailmikhaylov** [comment]
PTAL

**Comment by alexander-yevsyukov** [comment]
LGTM

**Pull request merged**